src/extended_type_info.cpp

```diff
@@ -125,7 +125,6 @@ BOOST_SERIALIZATION_DECL void
 extended_type_info::key_unregister() const{
     if(NULL == get_key())
         return;
-    BOOST_ASSERT(! singleton<detail::ktmap>::is_destroyed());
```

Comment:
The original author was suggesting that if the singleton is destroyed here, then something was quite wrong.  An invariant was broken somehow.  I agree with the original author on this one.  Why would the map have been destroyed AND something is calling unregister?  It sounds like there's a static uninitialization race somewhere.  Isn't this assertion still valid?  If not, there's still a race.

Reply:
Yes there is a broken invariant: With shared libs using static boost as described in the other occurence you commented on. Cause is simply implementation defined behaviour for shared libs (see difference Linux vs Windows and OSX)

So no the assertion is not valid anymore hence its removal. This condition is instead caught by the runtime check 1 line below.

Its is not a race as the destruction happens on unload in single-threaded way.

Reply:
Hmmm - as the original author, I think my intention is pretty clear.  So far I haven't managed to trip this assertion.  If does occur then I'd like to know about it.  Now I'm aware that you've got some tests which which do trip this assertion.  When I've looked at your tests before I've not been convinced that they prove what you want to prove.  Having said that, I'm willing to look at more tests.  But I would like to see them more carefully constructed.  I'm aware that this requires a lot more effort but I don't see any other choice.

Reply:
> So far I haven't managed to trip this assertion.

That's because `is_destroyed` doesn't work.

Reply:
Hmmm - how does it not work.
`
    static bool & get_is_destroyed(){
        static bool is_destroyed;
        return is_destroyed;
    }
public:
    BOOST_DLLEXPORT static T & get_mutable_instance(){
        BOOST_ASSERT(! get_singleton_module().is_locked());
        return get_instance();
    }
    BOOST_DLLEXPORT static const T & get_const_instance(){
        return get_instance();
    }
    BOOST_DLLEXPORT static bool is_destroyed(){
        return get_is_destroyed();
    }
    BOOST_DLLEXPORT singleton(){
        get_is_destroyed() = false;
    }
    BOOST_DLLEXPORT ~singleton() {
        get_is_destroyed() = true;
    }
`

where is the problem here?  I understand that the singleton destructor might not ever be called as it would happen post main time.  But that's not in anyone's control.  I don't know if one would say that that's an error.  In any case we don't depend on that and have no control over it anyway. The intention is to detect that case where some singleton is being uses after it's destructor has been called.  So to say it "doesn't work" would mean that ~singleton has been called but the is_destroyed still returns true.  I don't see any evidence of that happening.  Of course if assertions are being invoked - it would mean that singleton destructors are not being called in reverse sequence of the constructors.  The library is predicated on the assumption that does not happen.  Soooo, if this assertion is being invoked I would say that it's not an indication that the is_destroyed "doesn't work", it's more likely that the assumption of sequence of destruction/construction is not respected.  Of course this could be the scenario which might occur when one is invoking singletons from multiple dynamically loaded shared libraries.  Which I believe is the situtation we are trying to deal with in this case.  If this is indeed occurring - then the true question is - under what conditions does this occur and how can it be prevented.  

In other words, I believe that if the assertions are being invoked - that's it's a true violation of some assumption that I've made.  Not necessarily that the conditional is wrong.

Reply:
The constructor and destructor of `singleton<detail::ktmap>` are never called, because a variable of such type is never created.

Reply:
> I understand that the singleton destructor might not ever be called as it would happen post main time.

That's where you are mislead: The whole situation is about what happens "post main time": When all the singletons gets destructed and types get unregistered a mixup occurs and things go BOOM.

I described the test I did in #129 here: https://github.com/boostorg/serialization/pull/131#discussion_r232009933. I argue that this is a very carefully constructed test as it completely relies on the C++ standard. It shows under what circumstances the problem (`is_destroyed` returns false for a destroyed singleton) occurs: When the class is not inherited from singleton<> as it is the case for `singleton<detail::ktmap>` as @pdimov  also noticed. (There are 2 tests: 1 where it is inherited and 1 where it is not, both are valid as per your docu and code) This is why the subclass `singleton_wrapper` is required: To listen to the destruction of `detail::ktmap`, not (only) `singleton<>`

> If this is indeed occurring - then the true question is - under what conditions does this occur and how can it be prevented.

The first part can be verified easily: Cherry-pick https://github.com/boostorg/serialization/pull/131/commits/7a34ef8b76e542560ac3895b5a3d68499f7ed422 and run the test #128 on linux in debug mode: The assertion **will** fire. I do not think this can be prevented because it is "implementation defined land" (shared libs), but it can be gracefully handled (by the runtime checks)

Side note: What do you think is wrong with #128? I made this out of #111 (which is the minimum working example) with completely documented code. No tricks or anything. Or did I miss anything?

Reply:
>The constructor and destructor of singleton<detail::ktmap> are never called, because a variable of such type is never created.
Why do you think that? 
a) Certainly the get_instance which creates the singleton<detail::ktmap> is getting called.  Otherwise things would not work at all.  So the constructor must be getting called.  And there is a static instance created at pre-compile time so the constructor must be getting called there.
b) I can't guarantee that the destructor is getting at post-main time as I don't have much access to it.
c) Note that there is a test called "test_singleton" which verifies that singleton constructors are called and that destructors are called in reverse order.  See https://github.com/boostorg/serialization/blob/develop/test/test_singleton.cpp

Now the question arises as to what happens when we've got multiple singletons of the same type. Such as in different shared libraries.  It might be possible that there would be some sort of mixup in that results in the wrong destructor being called.  But I don't see an easy way to track this down.

Reply:
>That's where you are mislead: The whole situation is about what happens "post main time": When all the singletons gets destructed and types get unregistered a mixup occurs and things go BOOM.

That's conceivable. I'll take another look at your test.  I'll look at "test_plain" as it seems to be simpler.

>The first part can be verified easily: Cherry-pick 7a34ef8 and run the test #128 on linux in debug mode: The assertion will fire. I do not think this can be prevented because it is "implementation defined land" (shared libs), but it can be gracefully handled (by the runtime checks)

Right - I don't have a linux system setup up for this.  But I'll still review the test.  But I want to review the best test so if you think there's a better one than #129/test_plain let me know.

If your assessment of the problem above is correct, it should manifest itself replacing singleton<detail::ktmap> with some simpler type - Correct.  The reason I ask this is that I'd like to verify that it's an issue with singleton and something else in the serialization library.  Perhaps you'd like to make that change and report the results.

If you assessment above is correct, then we need a workaround.  The workaround is not the complex in theory - just suppress the assertion at post main time. I'm sure a way to do that could be found.  But of course everything is trickier than it looks.  In particular, what is post-main time when a dll is unloaded?  Is it really post main time when things go south - or is it at dll unload time? etc. etc....

Reply:
> Certainly the `get_instance` which creates the `singleton<detail::ktmap>` is getting called. Otherwise things would not work at all. So the constructor must be getting called.

`singleton<detail::ktmap>::get_instance` is a static member. It creates an instance of `detail::ktmap` and returns it. This however has nothing to do with the constructor of `singleton<detail::ktmap>`.

Reply:
In the case where `X` derives from `singleton<X>`, constructing a `X` necessarily constructs the base `singleton<X>`. `detail::ktmap`, however, is a typedef for `std::multimap<>`, which does not derive from `singleton<>`.

This is all explained in Alexander's test which checks both cases (he calls them "plain" singleton and "inherited" singleton.)

Reply:
> This however has nothing to do with the constructor of singleton<detail::ktmap>

Hmmm - does this not invoke the constructor of singleton<detail::ktmap> at pre-main time?

template<class T>
T & singleton< T >::m_instance = singleton< T >::get_instance();

Reply:
No; `T` here is `detail::ktmap`, so it defines a variable of type `detail::ktmap&`, initialized with the result of calling `singleton<detail::ktmap>::get_instance()`. Nothing here constructs a `singleton<detail::ktmap>`.

Reply:
OK - I see this now and it's not what I intended.  I'll look into this

Reply:
OK - I'm buying this.  I'll put in my current version and run some tests on my machine then check it in.  I did change the m_instance from a reference to a pointer to avoid an error with UBSAN set.  Hopefully that won't break anything.

Reply:
@pdimov already said the vital parts but I still want to answer the following to avoid future problems:

> But I want to review the best test so if you think there's a better one than #129/test_plain let me know.

Actually this is answered in the description of #129: There are 2 possible use cases for singleton. `extended_type_info` uses the inherited-paradigm and `tkmap` the plain-paradigm. For both `is_destroyed` has to work. That's why **both**  tests are required. So please do not only merge one of them or it will be broken again.